src/Microsoft.ML.Data/MLContext.cs

```diff
@@ -61,14 +63,16 @@ public sealed class MLContext : IHostEnvironment
         /// </summary>
         public Action<string> Log { get; set; }
 
+        public AggregateCatalog PartCatalog { get; }
```

Comment:
I'm wondering how this design would play in a larger application that already uses MEF.  If I was writing the application, would I want my "ML.NET MEF stuff" separate from the rest of my application's "MEF stuff"?  Or would I want a unified `CompositionContainer` that contained all my components.

Say my `CustomMappingTransformer` `Imports` a bunch of other stuff in my application. I would need to duplicate all those "parts" into this catalog.

What if we changed the design such that a consumer supplies the `CompositionContainer` to the `MLContext` instead? They can configure it any way they want.

Reply:
Notes:

I haven't ever done this myself, so I'm not sure on the best practices in this area. The above idea was just me brainstorming - not really a "I believe this is the best way to accomplish this design".

Maybe @weshaggard might have some ideas on what would be the best way to expose this to users?

Reply:
Actually I like this idea better. Note that I mutate the container sometimes though...

---
In reply to: [229440195](https://github.com/dotnet/machinelearning/pull/1406#discussion_r229440195) [](ancestors = 229440195)

Reply:
You (i.e. ML.NET) shouldn't have to mutate the container, should it?  If the user needs/wants to `[Import] MLContext`, then they can add that to the container themselves, can't they?  (or am I missing something?)

Reply:
They are free to do this if they want. I thought it's nice not to have to do it though. Essentially, they will ALWAYS need an instance of `MLContext` to instantiate any `ITransformer`, and if we were to do it for them optionally, I think it should be fine?

---
In reply to: [229446216](https://github.com/dotnet/machinelearning/pull/1406#discussion_r229446216) [](ancestors = 229446216)

Reply:
Sounds fine to me.